app/Libraries/Mods.php

```diff
@@ -0,0 +1,211 @@
+<?php
+
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+declare(strict_types=1);
+
+namespace App\Libraries;
+
+use App\Exceptions\InvariantException;
+use Ds\Set;
+
+class Mods
+{
+    // A => B: mod A implies mod B.
+    const IMPLIED_MODS = [
+        'NC' => 'DT',
+        'PF' => 'SD',
+    ];
+
+    const LEGACY_BITSET = [
+        1 << 0 => 'NF' ,
+        1 << 1 => 'EZ' ,
+        1 << 2 => 'TD' ,
+        1 << 3 => 'HD' ,
+        1 << 4 => 'HR' ,
+        1 << 5 => 'SD' ,
+        1 << 6 => 'DT' ,
+        1 << 7 => 'RX' ,
+        1 << 8 => 'HT' ,
+        1 << 9 => 'NC' ,
+        1 << 10 => 'FL' ,
+        1 << 12 => 'SO' ,
+        1 << 13 => 'AP' ,
+        1 << 14 => 'PF' ,
+        1 << 20 => 'FI' ,
+        1 << 29 => 'V2' ,
+        1 << 30 => 'MR' ,
+
+        // mania keys (converts)
+        1 << 15 => '4K' ,
+        1 << 16 => '5K' ,
+        1 << 17 => '6K' ,
+        1 << 18 => '7K' ,
+        1 << 19 => '8K' ,
+        1 << 24 => '9K' ,
+    ];
+
+    const LEGACY_PREFERENCE_MODS_BITSET = 0b01000000000000000100001000100000; // SD, NC, PF, MR
+
+    public array $idToBitsetMap;
+    public Set $difficultyReductionIds;
+    public array $mods = [];
+
+    public function __construct()
+    {
+        $this->idToBitsetMap = array_flip(static::LEGACY_BITSET);
+        $this->difficultyReductionIds = new Set();
+
+        $metadata = json_decode(file_get_contents(database_path('mods.json')), true);
```

Comment:
Does this somehow get cached?

Reply:
yes, the class is initialised as a singleton

Reply:
yeah